File: etl/scripts/pullback_config.py
"""눌림목 전략 설정 로더. 필수 JSON이 없거나 잘못되면 주문을 중단한다."""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


REQUIRED_KEYS = {
    "budget_per_stock", "max_new_positions", "tp", "sl",
    "max_wait_days", "max_hold_days",
}


def path() -> Path:
    return Path(__file__).resolve().parent / "pullback.json"


def _integer(value: Any, name: str, minimum: int, maximum: int | None = None) -> None:
    if isinstance(value, bool) or not isinstance(value, int) or value < minimum:
        raise ValueError(f"{name} must be an integer >= {minimum}")
    if maximum is not None and value > maximum:
        raise ValueError(f"{name} must be <= {maximum}")


def validate(config: dict[str, Any]) -> None:
    missing = REQUIRED_KEYS - set(config)
    if missing:
        raise ValueError(f"pullback config missing keys: {', '.join(sorted(missing))}")
    _integer(config["budget_per_stock"], "budget_per_stock", 1)
    _integer(config["max_new_positions"], "max_new_positions", 1, 3)
    _integer(config["max_wait_days"], "max_wait_days", 1, 5)
    _integer(config["max_hold_days"], "max_hold_days", 1, 5)
    for name in ("tp", "sl"):
        value = config[name]
        if isinstance(value, bool) or not isinstance(value, (int, float)) or not 0 < value <= 1:
            raise ValueError(f"{name} must be a number > 0 and <= 1")


def load(config_path: Path | None = None) -> dict[str, Any]:
    target = config_path or path()
    config = json.loads(target.read_text(encoding="utf-8"))
    if not isinstance(config, dict):
        raise ValueError("pullback config must be a JSON object")
    validate(config)
    return config

File: broker-web/lib/pullback-config.ts
import fs from "node:fs/promises";
import path from "node:path";

export type PullbackConfig = {
  budget_per_stock: number;
  max_new_positions: number;
  tp: number;
  sl: number;
  max_wait_days: number;
  max_hold_days: number;
};

function repoRoot() {
  return path.join(process.cwd(), "..");
}

export const PATH = () => path.join(repoRoot(), "etl", "scripts", "pullback.json");

function integer(value: number, name: string, minimum: number, maximum?: number) {
  if (!Number.isInteger(value) || value < minimum || (maximum !== undefined && value > maximum)) {
    throw new Error(`${name} 값이 허용 범위를 벗어났습니다`);
  }
}

export function validate(config: PullbackConfig): void {
  if (!config || typeof config !== "object") throw new Error("설정 형식이 올바르지 않습니다");
  integer(config.budget_per_stock, "종목당 예산", 1);
  integer(config.max_new_positions, "하루 최대 신규 매수", 1, 3);
  integer(config.max_wait_days, "관찰 거래일", 1, 5);
  integer(config.max_hold_days, "최대 보유 거래일", 1, 5);
  for (const [key, label] of [["tp", "익절"], ["sl", "손절"]] as const) {
    const value = config[key];
    if (typeof value !== "number" || !Number.isFinite(value) || value <= 0 || value > 1) {
      throw new Error(`${label}은 0% 초과 100% 이하여야 합니다`);
    }
  }
}

export async function read(target = PATH()): Promise<PullbackConfig> {
  const config = JSON.parse(await fs.readFile(target, "utf-8")) as PullbackConfig;
  validate(config);
  return config;
}

export async function write(config: PullbackConfig, target = PATH()): Promise<void> {
  validate(config);
  const previous = await fs.readFile(target, "utf-8");
  await fs.writeFile(target + ".bak", previous, "utf-8");
  await fs.writeFile(target, JSON.stringify(config, null, 2) + "\n", "utf-8");
}
